torch/csrc/autograd/record_function.h

```diff
@@ -26,31 +28,79 @@ struct TORCH_API StringView {
   const char* str_ptr_;
 };
 
-using GetPackedInputsCallback = std::function<std::vector<c10::IValue>()>;
+enum class RecordFunctionState {
+  NO_CALLBACKS, // no callbacks registered, no need to use RecordFunction
+  NAME_ONLY, // callbacks need only basic information (function name)
+  NAME_AND_INPUTS, // callbacks need both name and function inputs
+};
+TORCH_API RecordFunctionState recordFunctionState();
 
 struct TORCH_API RecordFunction {
-  explicit RecordFunction(Function* fn, GetPackedInputsCallback cb = nullptr);
+  // Default constructor is used with before function called afterwards
```

Comment:
There are now 3 constructors, and 7 before functions! I can't believe it has to be this complicated. Do we need any constructor but the no-arg one? Why is anyone constructing RecordFunction directly after this patch? At least 3/7 before functions are just to forward inputs. But adding the inputs needs to be a separate function call anyway so that it can have guarded execution.

Reply:
Sure, if you think we could restrict ourselves to `before` usage then we can remove non-trivial constructors, they serve a shortcut purpose to avoid explicit `before` usage